pinot-core/src/main/java/org/apache/pinot/core/data/manager/BaseTableDataManager.java

```diff
@@ -1013,8 +1013,30 @@ public boolean tryLoadExistingSegment(SegmentZKMetadata zkMetadata, IndexLoading
     }
   }
 
+  /**
+   * This method checks if any reload is needed on a segment.
+   * Scans through indices, columns, startree index and min max values to achieve the result
+   * True if reload is needed and false if no reload is needed.
+   */
+  @Override
+  public boolean checkReloadSegment(SegmentZKMetadata zkMetadata, IndexLoadingConfig indexLoadingConfig) {
+    String segmentName = zkMetadata.getSegmentName();
+    SegmentDirectory segmentDirectory =
+        tryInitSegmentDirectory(segmentName, String.valueOf(zkMetadata.getCrc()), indexLoadingConfig);
+
+    try {
+      Schema schema = indexLoadingConfig.getSchema();
+      //if re processing or reload is needed on a segment then return true
+      return ImmutableSegmentLoader.needPreprocess(segmentDirectory, indexLoadingConfig, schema);
+    } catch (Exception e) {
+      _logger.error("Failed to check if reload is needed for a segment {}", segmentName, e);
+      closeSegmentDirectoryQuietly(segmentDirectory);
+      return false;
```

Comment:
I saw you stated in the API description that `500` will be returned if there is any server side error. Shouldn't we return `500` here in this case?